### PR TITLE
Complete migration of `Scalar` to functions

### PR DESCRIPTION
**Pull request by james-pre**
Back in February, I opened #14805. One of the examples I used was `Scalar`. @deltakosh started to make the migration to functions in https://github.com/BabylonJS/Babylon.js/commit/74c4c4740dde6488a8dd46d2a50917a6cf83a06f.

This PR completes the migration from static methods on `Scalar` to functions. I've updated all usage throughout the framework, and marked the original class as deprecated, though I have not removed it for backward compatibility.

It also renames the functions that are named using pascal case to camel case in order to follow the correct convention. Again, they were not removed for backward compatibility but have been marked as deprecated.

One notable change is that I omitted the `Scalar.Log2` and `Scalar.Sign` method. This is because `Math.log2` and `Math.sign` can be used instead, which should be faster since it is a native function.

Using functions should speed up performance and reduce bundle size.

If pascal case is desired for the function names, I'm flexible and can make the change in a minute so no worries there.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Review by deltakosh — APPROVED** [review]
Appreciated!

**Comment by sebavan** [comment]
@RaananW anyway to prevent reintroducing the old one ?

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Review by sebavan — CHANGES_REQUESTED** [review]
@james-pre You are introducing breaking changes AGAIN in naming of the functions

File: packages/dev/core/src/Maths/math.scalar.functions.ts [hunk 1]
```diff
@@ -38,10 +38,78 @@ export function RandomRange(min: number, max: number): number {
  * @param amount amount to lerp between
  * @returns the lerped value
  */
-export function Lerp(start: number, end: number, amount: number): number {
+export function lerp(start: number, end: number, amount: number): number {
```
Comment (deltakosh) [inline packages/dev/core/src/Maths/math.scalar.functions.ts:41]: We use a capital for exported functions

**Review by deltakosh — CHANGES_REQUESTED** [review]

**Comment by deltakosh** [comment]
pascal case is Actually mandatory. You need to keep in mind we dearly protect back compat

File: packages/dev/core/src/Maths/math.scalar.functions.ts (hunk 1 above)
Comment (james-pre) [reply packages/dev/core/src/Maths/math.scalar.functions.ts]: ede6de3

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by james-pre** [comment]
> You are introducing breaking changes AGAIN in naming of the functions

@sebavan, My bad, I thought we were moving to normal/standard naming conventions for functions. I reverted the naming change and will keep the pascal case convention in the future.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by RaananW** [comment]
> @RaananW anyway to prevent reintroducing the old one ?

Not sure what you mean. The Scalar object? No way around it. We wither export everything "as", or create an object (that will be tree-shaken out in es6) that can be exported. As people might have already imported it from this path we also can't move it to a new place (like index.ts, where it makes more sense to have it).

**Comment by james-pre** [comment]
Just FYI, I changed `Scalar` to a normal object instead of a class, and used some ES6 destructing syntax to avoid repeating all of the function names. This also copies all of the documentation, so overall [ad60a82](https://github.com/BabylonJS/Babylon.js/pull/15373/commits/ad60a82f22a0300ee7afe17ec3f9c44f2bd52004) has a lot of de-duping.

**Comment by james-pre** [comment]
CI needs a rerun because Node failed to download.

**Comment by RaananW** [comment]
> CI needs a rerun because Node failed to download.

Done

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by james-pre** [comment]
Build for `ad60a82` failed because `../babylonjs/babylon.module.d.ts(187481,21): error TS2552: Cannot find name 'functions'. Did you mean 'Function'?`

This was because math.scalar.ts looks like this (minus comments):
```ts
import * as functions from "./math.scalar.functions";

export const Scalar = {
    ...functions,
    TwoPi: Math.PI * 2,
    Sign: Math.sign,
    Log2: Math.log2,
    HCF: functions.HighestCommonFactor, // <- error occurs here
};
```

`...functions` gets expanded in the generated types, but not `function.HighestCommonFactor`.

This results in the generated `Scalar`type being:
```ts
    export var Scalar: {

        TwoPi: number;
        Sign: (x: number) => number;
        Log2: {
            (x: number): number;
            (x: number): number;
        };
        HCF: typeof functions.HighestCommonFactor; // <- functions does not exist
        ExtractAsInt(value: number): number;
        WithinEpsilon(a: number, b: number, epsilon?: number): boolean;
        // ... rest of functions
    };
```

Which results in a type error.

I tried to fix this by adding a named `HCF` export for `HighestCommonFactor` (it's marked as deprecated), and then removing the alias in `Scalar`, but this has not fixed the issue.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by james-pre** [comment]
fa2046b Should fix the problem. By using a getter, the type can be manually specified. It looks good in my local build.

Also in fa2046b, I changed the scalar.functions.ts export in Maths/index.ts to a namespace export (`scalar`), which prevents some crowding of the global scope.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by james-pre** [comment]
There are some issues with the types.

Looking at the first error:
```
../babylonjs-gui/babylon.gui.module.d.ts(9327,9): error TS2416: Property 'getAlphaTestTexture' in type 'MRDLSliderThumbMaterial' is not assignable to the same property in base type 'PushMaterial'.
  Type '() => BABYLON.Nullable<BABYLON.BaseTexture>' is not assignable to type '() => BABYLON.Nullable<BaseTexture>'.
    Type 'BABYLON.Nullable<BABYLON.BaseTexture>' is not assignable to type 'BABYLON.Nullable<BaseTexture>'.
      Type 'BaseTexture' is missing the following properties from type 'BaseTexture': sphericalPolynomial, forceSphericalPolynomialsRecompute
```

I believe it is because of this interface merging:
```ts
declare module "./baseTexture" {
    export interface BaseTexture {
        sphericalPolynomial: Nullable<SphericalPolynomial>;
        forceSphericalPolynomialsRecompute(): void;
    }
}
```
[core/Materials/Textures/baseTexture.polynomial.ts](https://github.com/BabylonJS/Babylon.js/blob/d61c9c1f4cacbe155e1506875dd21c6748cadf69/packages/dev/core/src/Materials/Textures/baseTexture.polynomial.ts)

So, it looks like the type for `BABYLON.BaseTexture` does not get the interface merging while `BaseTexture` does?

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Maths/index.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 export * from "./math.scalar";
-export * from "./math.functions";
+export * as scalar from "./math.functions";
```
Comment (sebavan) [inline packages/dev/core/src/Maths/index.ts:2]: Why this change ? I do not think it should be here

**Review by sebavan — DISMISSED** [review]

File: packages/dev/core/src/Maths/index.ts (hunk 2 above)
Comment (james-pre) [reply packages/dev/core/src/Maths/index.ts:2]: I did this to avoid polluting the `BABYLON` scope. I can revert this if you'd like?

Comment (sebavan) [reply packages/dev/core/src/Maths/index.ts:2]: yup as it is a breaking change for the functions which were already there, thanks

**Comment by sebavan** [comment]
cc @RaananW for the UMD build issue in case he has an easy way to triple check the root cause.

**Comment by RaananW** [comment]
i'll check why build fails here. probably some declaration processing. Will keep you updated.

File: packages/dev/core/src/Maths/index.ts (hunk 2 above)
Comment (RaananW) [reply packages/dev/core/src/Maths/index.ts:2]: The line is not needed. in UMD the Scalar object should still be present.

Comment (RaananW) [reply packages/dev/core/src/Maths/index.ts:2]: This is also the reason the build was failing.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Maths/index.ts (hunk 2 above)
Comment (james-pre) [reply packages/dev/core/src/Maths/index.ts]: Fixed in https://github.com/BabylonJS/Babylon.js/pull/15373/commits/df7e2b5b17138058e718aace2bdd5421dae67623.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15373/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15373/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15373/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15373/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15373/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15373/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/testResults/webgl2playwright/index.html

**Comment by RaananW** [comment]
I know what the issue here is, but I won't be able to solve it today. Will tackle this on Monday.

**Comment by james-pre** [comment]
> I know what the issue here is, but I won't be able to solve it today. Will tackle this on Monday.

Thanks @RaananW! I really appreciate all the time you've put into this kind of stuff recently. 

I may be preoccupied next week since I will be moving, but I'll try to make time for this.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15373/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by james-pre** [comment]
Thanks, I wasn't sure why tests were failing, so I appreciate your expertise. Thanks again.

**Review by sebavan — APPROVED** [review]
Any way to add a rule to never add import from the wrong file internally ?

**Comment by RaananW** [comment]
> Any way to add a rule to never add import from the wrong file internally ?

it will fail in CI, that's a wonderful rule :-)

And no, it is a valid import when done in the mono-repo. it is only in UMD that we get to see this problem.
